hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCompactionChore.java

```diff
@@ -0,0 +1,253 @@
+            }
+          } catch (IOException e) {
+            String errMsg = String.format("Failed to compact table=%s cf=%s",
+              htd.getTableName(), hcd.getNameAsString());
+            LOG.error(errMsg, e);
+          } catch (InterruptedException ee) {
+            Thread.currentThread().interrupt();
+            master.reportMobCompactionEnd(htd.getTableName());
+            String warnMsg = String.format("Failed to compact table=%s cf=%s",
+              htd.getTableName(), hcd.getNameAsString());
+            LOG.warn(warnMsg, ee);
+            // Quit the chore
+            return;
+          }
+        }
+        if (reported) {
+          master.reportMobCompactionEnd(htd.getTableName());
+          reported = false;
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to compact", e);
+    }
+  }
+
+  @VisibleForTesting
+  public void performMajorCompactionInBatches(Admin admin, TableDescriptor htd,
+      ColumnFamilyDescriptor hcd) throws IOException, InterruptedException {
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      LOG.debug(
+        "Table={} cf={} - performing major MOB compaction in non-batched mode,"
+            + "regions={}, batch size={}",
+        htd.getTableName(), hcd.getNameAsString(), regions.size(), regionBatchSize);
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>(regions.size());
+    for (int i = 0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for (RegionInfo ri : toCompact) {
+      startCompaction(admin, htd.getTableName(), ri, hcd.getName());
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>(toCompact.size());
+    List<RegionInfo> failed = new ArrayList<RegionInfo>();
+    int totalCompacted = 0;
+    while (!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri : toCompact) {
+        try {
+          if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
+            totalCompacted++;
+            LOG.info(
+              "Finished major MOB compaction: table={} cf={} region={}," + " compacted regions={}",
+              htd.getTableName(), hcd.getNameAsString(), ri.getRegionNameAsString(),
+              totalCompacted);
+            compacted.add(ri);
+          }
+        } catch (IOException e) {
+          LOG.error("Could not get compaction state for table={} cf={} region={}, compaction will"+
+            " aborted for the region.",
+            htd.getTableName(), hcd.getNameAsString(), ri.getEncodedName());
+          LOG.error("Because of:", e);
```

Comment:
org.slf4j.Logger does not provide such API to have both: parametrized message and  Throwable object in a same call. At least, I did not manage to find it.

Reply:
the slf4j Logger methods for messages take vararg to do matching to parameters and also have special handling so that the very last parameter can be an exception.

ref: http://www.slf4j.org/faq.html#paramException